middleware_test.go

```diff
@@ -0,0 +1,252 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+* This file may have been modified by CloudWeGo authors. All CloudWeGo
+* Modifications are Copyright 2022 CloudWeGo Authors.
+*/
+
+package gzip
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/cloudwego/hertz/pkg/app"
+	"github.com/cloudwego/hertz/pkg/app/client"
+	"github.com/cloudwego/hertz/pkg/app/server"
+	"github.com/cloudwego/hertz/pkg/protocol"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGzipForClient(t *testing.T) {
+	h := server.Default(server.WithHostPorts("127.0.0.1:2333"))
+
+	h.GET("/ping", func(ctx context.Context, c *app.RequestContext) {
+		c.Header("Content-Length", strconv.Itoa(len(testResponse)))
+		c.String(200, testResponse)
+	})
+	go h.Spin()
+	time.Sleep(time.Second)
+
+	cli, err := client.NewClient()
+	if err != nil {
+		panic(err)
+	}
+	cli.Use(GzipForClient(DefaultCompression))
+
+	req := protocol.AcquireRequest()
+	res := protocol.AcquireResponse()
+
+	req.SetBodyString("bar")
+	req.SetRequestURI("http://127.0.0.1:2333/ping")
+
+	cli.Do(context.Background(), req, res)
+	if err != nil {
+		t.Fatalf("Get: %v", err)
+	}
+
+	assert.Equal(t, res.StatusCode(), 200)
+	assert.Equal(t, req.Header.Get("Vary"), "Accept-Encoding")
+	assert.Equal(t, req.Header.Get("Content-Encoding"), "gzip")
+	assert.NotEqual(t, req.Header.Get("Content-Length"), "0")
+	assert.NotEqual(t, fmt.Sprint(len(req.Body())), req.Header.Get("Content-Length"))
+}
+
+func TestGzipPNGForClient(t *testing.T) {
+	h := server.Default(server.WithHostPorts("127.0.0.1:2334"))
+
+	h.GET("/image.png", func(ctx context.Context, c *app.RequestContext) {
+		c.Header("Content-Length", strconv.Itoa(len(testResponse)))
+		c.String(200, testResponse)
+	})
+	go h.Spin()
+	time.Sleep(time.Second)
+
+	cli, err := client.NewClient()
+	if err != nil {
+		panic(err)
+	}
+	cli.Use(GzipForClient(DefaultCompression, WithExcludedExtensions([]string{".png"})))
+
+	req := protocol.AcquireRequest()
+	res := protocol.AcquireResponse()
+
+	req.SetBodyString("bar")
+	req.SetRequestURI("http://127.0.0.1:2334/image.png")
+
+	cli.Do(context.Background(), req, res)
+	if err != nil {
+		t.Fatalf("Get: %v", err)
+	}
+
+	assert.Equal(t, res.StatusCode(), 200)
+	assert.Equal(t, req.Header.Get("Vary"), "")
+	assert.Equal(t, req.Header.Get("Content-Encoding"), "")
+}
+
+func TestExcludedExtensionsForClient(t *testing.T) {
+	h := server.Default(server.WithHostPorts("127.0.0.1:3333"))
+
+	h.GET("/index.html", func(ctx context.Context, c *app.RequestContext) {
```

Comment:
Does the gzip client middleware work with the gzip server middleware, I didn't find it in the example or in the single test